File: sdk-ts/src/adapters/anthropic.ts
/**
 * Adapter for Claude tool use. No dependency on the `@anthropic-ai/sdk`
 * package — accepts either its SDK `ToolUseBlock` objects or the equivalent
 * raw object content block, since both shapes are common depending on how a
 * caller consumes the API.
 */
import type { Guard } from "../guard.ts";
import { getField } from "./compat.ts";

export async function dispatch(guard: Guard, toolUseBlock: unknown, handlers: Record<string, (args: unknown) => unknown>): Promise<unknown> {
  const name = getField<string>(toolUseBlock, "name");
  if (!name) {
    throw new Error(`tool_use_block has no 'name': ${JSON.stringify(toolUseBlock)}`);
  }
  const handler = handlers[name];
  if (!handler) {
    throw new Error(`no handler registered for tool '${name}'`);
  }

  const args = getField<Record<string, unknown>>(toolUseBlock, "input") ?? {};

  return guard.checkAndExecute(name, args, handler);
}

File: sdk-python/agentguard/adapters/anthropic.py
"""Adapter for Claude tool use. No dependency on the `anthropic` package —
accepts either its SDK `ToolUseBlock` objects or the equivalent raw dict
content block, since both shapes are common depending on how a caller
consumes the API.
"""
from __future__ import annotations

from typing import Any, Callable, Dict

from ..guard import Guard
from ._compat import get_field


def dispatch(guard: Guard, tool_use_block: Any, handlers: Dict[str, Callable[[dict], Any]]) -> Any:
    """Given one `tool_use` content block from a Claude response (an SDK
    `ToolUseBlock` or the equivalent raw dict) and a {tool_name: handler}
    map, policy-check the call and invoke the matching handler with its
    input if allowed.
    """
    name = get_field(tool_use_block, "name")
    if not name:
        raise ValueError(f"tool_use_block has no 'name': {tool_use_block!r}")
    if name not in handlers:
        raise KeyError(f"no handler registered for tool {name!r}")

    args = get_field(tool_use_block, "input") or {}

    return guard.check_and_execute(name, args, execute_fn=handlers[name])
